### PR TITLE
Add support for half-duplex SPI to CPy

### PR DESCRIPTION
**Pull request by sgauche**
I have a need for half-duplex SPI on a custom board.  I thought it would be worth adding to CircuitPython too.  I have tested this on the hardware and it is working.

### CODE REVIEW

**Review by tannewt — DISMISSED** [review]
I think it'd be good to add an explicit way to setup half duplex and document it in the matching portion of shared-bindings. I think it's best to provide an explicit half duplex bool. I'm worried that the majority of folks do want read to mean read from MISO and removing the exception will make it less clear what their coding error is.

It crossed my mind to make mosi == miso mean half duplex but I suspect that'd be a common copy and paste bug.

**Comment by sgauche** [comment]
I did think about that... instead of using if MISO and MOSI are defined, just let the user specify SPI_DIRECTION_2LINES_RXONLY, SPI_DIRECTION_2LINES, or SPI_DIRECTION_1LINE via a parameter when calling spi configure like:
`spi.configure(baudrate=5000000, com_mode=[full, half, simplex], phase=0, polarity=0)`

Would something like that be possible?

There are 3 modes, full-duplex (normal 4-wire SPI), half-duplex (3-wire SPI with Bidirectional data, and simplex (3-wire SPI with unidirectional data, RX only). Cpy was already set up to auto set SPI_DIRECTION_2LINES/full-duplex and SPI_DIRECTION_2LINES_RXONLY/simplex. So I don't think we can use a boolean, since there are 3 modes, not 2.

**Comment by sgauche** [comment]
I looked at that but only saw ways to pass bools and uints through the configure parameters, didn't see a way to pass strings.

**Comment by sgauche** [comment]
Now I'm realizing if we change the spi configure parameters, that touches all of the different ports, not just stm... hmm.

**Comment by dhalbert** [comment]
> Now I'm realizing if we change the spi configure parameters, that touches all of the different ports, not just stm... hmm.

Right, we need to change the `common_hal` routines even if it's not implemented.

I think a `half_duplex` boolean would be sufficient. You can complain if both pins are specified and `half_duplex` is true.

File: ports/atmel-samd/common-hal/busio/SPI.c [hunk 1]
```diff
@@ -57,6 +57,10 @@ void common_hal_busio_spi_construct(busio_spi_obj_t *self,
     uint8_t miso_pad = 0;
     uint8_t dopo = 255;
 
+    if (half_duplex == true) {
```
Comment (dhalbert) [inline ports/atmel-samd/common-hal/busio/SPI.c:60]:
```suggestion
    if (half_duplex) {
```
This is more canonical C.

**Review by dhalbert — CHANGES_REQUESTED** [review]
Thanks! Minor style request.

**Comment by sgauche** [comment]
I'm guessing the too little flash failures are because I added `mp_raise_NotImplementedError(translate("Half duplex SPI is not implemented"));`. I could just do a `mp_raise_NotImplementedError(Null);`?

**Comment by dhalbert** [comment]
Don't worry about these "too little flash" errors. I will push some commits that squeeze the builds that are overflowing. An unlabeled "NotImplementedError" is confusing. This happens all the time when we add functionality to tiny builds.

**Review by dhalbert — APPROVED** [review]
Thanks for your perseverance on this!

File: shared-bindings/busio/SPI.c [hunk 2]
```diff
@@ -74,7 +74,8 @@
 //|
 //|         :param ~microcontroller.Pin clock: the pin to use for the clock.
 //|         :param ~microcontroller.Pin MOSI: the Main Out Selected In pin.
-//|         :param ~microcontroller.Pin MISO: the Main In Selected Out pin."""
+//|         :param ~microcontroller.Pin MISO: the Main In Selected Out pin.
+//|         :param bool half_duplex: True when MOSI is used for bidirectional data. False when SPI is full-duplex or simplex."""
```
Comment (tannewt) [inline shared-bindings/busio/SPI.c:78]: Note that we should have a `half_duplex: bool = False` in the function signature.

Comment (sgauche) [reply shared-bindings/busio/SPI.c:78]: Where is that?

Comment (dhalbert) [reply shared-bindings/busio/SPI.c:78]: Address by #6003.

Comment (sgauche) [reply shared-bindings/busio/SPI.c:78]: Never mind, looks like dhalbert is on it

**Pull request merged**